### PR TITLE
replace reference parameter with std::pair

### PR DESCRIPTION
**Pull request by neheb**
Simpler and allows to use structured bindings.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

### CODE REVIEW

**Comment by neheb** [comment]
Actually... std::string can just be used.

**Comment by neheb** [comment]
Hmm that's probably not correct.

File: src/util/upnp_quirks.cc [hunk 1]
```diff
@@ -56,8 +56,8 @@ void Quirks::addCaptionInfo(const std::shared_ptr<CdsItem>& item, const std::uni
     if (!startswith(item->getMimeType(), "video"))
         return;
 
-    std::string url;
-    if (UpnpXMLBuilder::renderSubtitle(context->getServer()->getVirtualUrl(), item, url)) {
+    auto [url, r] = UpnpXMLBuilder::renderSubtitle(context->getServer()->getVirtualUrl(), item);
```
Comment (KarlStraussberger) [inline src/util/upnp_quirks.cc:59]: Please use `subAdded` instead of single character var

File: src/web/containers.cc [hunk 2]
```diff
@@ -74,8 +74,8 @@ void web::containers::process()
         int autoscanType = cont->getAutoscanType();
         ce.append_attribute("autoscan_type") = mapAutoscanType(autoscanType).data();
 
-        std::string url;
-        if (UpnpXMLBuilder::renderContainerImage(server->getVirtualUrl(), cont, url)) {
+        auto [url, r] = UpnpXMLBuilder::renderContainerImage(server->getVirtualUrl(), cont);
```
Comment (KarlStraussberger) [inline src/web/containers.cc:77]: Same here

File: src/web/edit_load.cc [hunk 3]
```diff
@@ -195,8 +195,8 @@ void web::edit_load::process()
         mimeType.append_attribute("value") = objItem->getMimeType().c_str();
         mimeType.append_attribute("editable") = true;
 
-        std::string url;
-        if (UpnpXMLBuilder::renderItemImage(server->getVirtualUrl(), objItem, url)) {
+        auto [url, r] = UpnpXMLBuilder::renderItemImage(server->getVirtualUrl(), objItem);
```
Comment (KarlStraussberger) [inline src/web/edit_load.cc:198]: Same

File: src/web/items.cc [hunk 4]
```diff
@@ -133,8 +133,8 @@ void web::items::process()
         std::string res = UpnpXMLBuilder::getFirstResourcePath(objItem);
         item.append_child("res").append_child(pugi::node_pcdata).set_value(res.c_str());
 
-        std::string url;
-        if (UpnpXMLBuilder::renderItemImage(server->getVirtualUrl(), objItem, url)) {
+        auto [url, r] = UpnpXMLBuilder::renderItemImage(server->getVirtualUrl(), objItem);
```
Comment (KarlStraussberger) [inline src/web/items.cc:136]: Same

**Comment by neheb** [comment]
I think this is what you meant.

**Comment by neheb** [comment]
To clarify, my original idea was to check if string is not empty, but not sure if that would break things.

**Comment by KarlStraussberger** [comment]
It's fine the way it is now. Never rely on particular values for control - separation of concerns

**Comment by neheb** [comment]
Interestingly enough, both commits have the same compiled size. Guess gcc sees through this and optimizes it.

**Comment by KarlStraussberger** [comment]
> 
> 
> Interestingly enough, both commits have the same compiled size. Guess gcc sees through this and optimizes it.

Sometimes it's just by accident. Logic and readability should always dominate over compiled sized, especially if we consider different compilers on different platforms. Our _work_ is logic - theirs to make small code out of it.

**Pull request merged**